tentacle/src/transports/mod.rs

```diff
@@ -28,6 +36,40 @@ pub trait Transport {
     fn dial(self, address: Multiaddr) -> Result<Self::DialFuture>;
 }
 
+pin_project! {
+    pub struct TransportFuture<T, F> {
+        #[pin]
+        executed: T,
+        maker: PhantomData<F>
+    }
+}
+
+impl<T, F> TransportFuture<T, F>
+where
+    T: Future<Output = Result<(Multiaddr, F)>> + 'static,
+{
+    pub fn new(executed: T) -> TransportFuture<T, F> {
+        TransportFuture {
+            executed,
+            maker: PhantomData::default(),
+        }
+    }
+}
```

Comment:
we can use a simple trait wrapper here:
```
pub struct TransportFuture<T> {
    executed: T,
}

impl<T> TransportFuture<T> {
    pub fn new(executed: T) -> TransportFuture<T> {
        TransportFuture { executed }
    }
}

impl<T> Future for TransportFuture<T> where T: Future {
    type Output = T::Output;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let executed = unsafe {
            let this = self.get_unchecked_mut();
            Pin::new_unchecked(&mut this.executed)
        };
        executed.poll(cx)
    }
}
```

Reply:
It is a little different in this way of writing. There is no effective constraint on `executed` when `new`. Since this type is provided by our own library, there is no problem with writing this way.